rllib/utils/replay_buffers/utils.py

```diff
@@ -52,11 +53,34 @@ def update_priorities_in_replay_buffer(
             #  policies (note: fixing this in torch_policy.py will
             #  break e.g. DDPPO!).
             td_error = info.get("td_error", info[LEARNER_STATS_KEY].get("td_error"))
+
+            policy_batch = train_batch.policy_batches[policy_id]
             # Set the get_interceptor to None in order to be able to access the numpy
             # arrays directly (instead of e.g. a torch array).
-            train_batch.policy_batches[policy_id].set_get_interceptor(None)
+            policy_batch.set_get_interceptor(None)
             # Get the replay buffer row indices that make up the `train_batch`.
-            batch_indices = train_batch.policy_batches[policy_id].get("batch_indexes")
+            batch_indices = policy_batch.get("batch_indexes")
+
+            if SampleBatch.SEQ_LENS in policy_batch:
+                # Batch_indices are represented per column, in order to update
+                # priorities, we need one index per td_error
+                _batch_indices = []
+
+                # Sequenced batches have been zero padded to max_seq_len.
+                # Depending on how batches are split during learning, not all
+                # sequences have an associated td_error (trailing ones missing).
+                if policy_batch.zero_padded:
+                    seq_lens = len(td_error) * [policy_batch.max_seq_len]
+                else:
+                    seq_lens = policy_batch[SampleBatch.SEQ_LENS][: len(td_error)]
```

Comment:
Just curious:
What would lead to a situation where:
```
policy_batch[SampleBatch.SEQ_LENS][: len(td_error)] != policy_batch[SampleBatch.SEQ_LENS]
```

Reply:
I was wondering thinking about this a lot because `len(td_error)` never seemed to fit the sequence lengths. Even if I took into account `max_seq_len` because of padding.

The problem in RNNSAC is: It uses SAC's train_batch_size of 256. While padding the sequence to prepare it for feeding it into an RNN, we increase the size (to 264 in this example). This blows up our sequences and we end up having an incomplete last sequence to train on because, before learning, we cut the batch to 256. Therefore the last sequence would always be incomplete.

<img width="770" alt="Screenshot 2022-08-09 at 18 04 22" src="https://user-images.githubusercontent.com/9356806/183701941-6ee26ecb-28df-4469-9318-19fe9baf909e.png">

In our Policies, we cut loaded batches if they are too long. For example this snippet is from our current TorchPolicyV2:
 ```
             if device_batch_size >= len(self._loaded_batches[0][0]):
                batch = self._loaded_batches[0][0]
            else:
                batch = self._loaded_batches[0][0][offset : offset + device_batch_size]
```

So if you were to learn on a batch that has sequences, but is not padded (for whatever reason), and is larger than the device_batch_size, the produced td_error would be smaller than the number of sequences. Therefore, the discarded sequences would not produce a TD error and so `len(td_error)!=len(batch.SEQ_LENS)`.